tianshou/data/batch.py

```diff
@@ -29,6 +29,30 @@ def _is_batch_set(data: Any) -> bool:
     return False
 
 
+def is_scalar_value(value: Any) -> bool:
+    # check if the value is a scalar
+    # 1. python bool object, number object: isinstance(value, Number)
+    # 2. numpy scalar: isinstance(value, np.generic)
+    # 3. python object rather than dict / Batch / tensor
+    # the check of dict / Batch is omitted because this only checks a value.
+    # a dict / Batch will eventually check their values
+    is_scalar = isinstance(value, Number)
+    is_scalar = is_scalar or isinstance(value, np.generic)
```

Comment:
why not to use np.asanyarray(value).size == 1 instead of anything else ? Is it really slower ?

Reply:
> why not to use np.asanyarray(value).size == 1 

That's a cool solution! Haven't thought about that yet.

Reply:
Btw, asanyarray is mentioned in reward scalar

Reply:
After a second thought, I still vote for the current implementation.

If we use ``np.asanyarray(value).size == 1``, additionally we have to check ``not np.asanyarray(value).shape`` to avoid ``[[1]]``

Furthermore, the current implementation makes it clear what types of data are considered scalars. But ``np.asanyarray(value).size == 1`` leaves all these to numpy and may confuse future maintainers.

Third, ``np.asanyarray`` does not come for free, though. It must actually does these type check, too.

Reply:
> If we use np.asanyarray(value).size == 1, additionally we have to check not np.asanyarray(value).shape to avoid [[1]]

Indeed, that's the point of this approach, containing a single value, same goes for any array with shape [1, 1, ...., 1] x)

> Furthermore, the current implementation makes it clear what types of data are considered scalars. But np.asanyarray(value).size == 1 leaves all these to numpy and may confuse future maintainers.

I don't think so. It just means: it must be possible to cast this think in a numpy array having a single element. It is pretty clear to me, knowing that it is in the name of the function itself and the docstring.

> Third, np.asanyarray does not come for free, though. It must actually does these type check, too.

Of course, there is no free lunch.

Reply:
If we use ``np.asanyarray(value).size`` to check whether it is a scalar, maybe we can take a step further, and just store the data with np.ndarray. This can remove many unnecessary type checks.

Current status:
```python
a = Batch(a=1)
type(a.a) # int
```

Proposed:
```python
a = Batch(a=1)
type(a.a) # np.ndarray
print(a.a) # array(1)
```

Reply:
> If we use np.asanyarray(value).size to check whether it is a scalar

I'm not saying we must do it though, just that it must be considered.

> maybe we can take a step further, and just store the data with np.ndarray. This can remove many unnecessary type checks.

Personally I don't see any reason not to convert scalars into numpy array systematically if it is actually un-noticeable for the user. Remove unnecessary type check if a good thing.